File: src/ipc/dto.ts
// DTOs - Data Transfer Objects
//
// Purpose: TypeScript interfaces for IPC communication between frontend and Tauri backend
//
// Keep in sync with Rust DTOs in backend

// ==================== COMMAND PAYLOADS ====================

// Project Commands
export interface ProjectNewPayload {
  name: string;
}

export interface ProjectSavePayload {
  path?: string; // If omitted, save to current path
}

export interface ProjectLoadPayload {
  path: string;
}

export interface ProjectExportPayload {
  path: string;
  format: "wav" | "ogg" | "mp3";
}

// Transport Commands
export interface TransportPlayPayload {
  position?: number; // Start position in beats
}

export interface TransportStopPayload {
  // No payload
}

export interface TransportPausePayload {
  // No payload
}

export interface TransportSeekPayload {
  position: number; // Position in beats
}

export interface TransportSetBpmPayload {
  bpm: number;
}

export interface TransportSetTimeSignaturePayload {
  numerator: number;
  denominator: number;
}

// Grid Commands
export interface GridSetCellPayload {
  channel: number;
  row: number;
  col: number;
  velocity: number;
  enabled: boolean;
}

export interface GridClearCellPayload {
  channel: number;
  row: number;
  col: number;
}

export interface GridClearChannelPayload {
  channel: number;
}

export interface GridClearAllPayload {
  // No payload
}

export interface GridCopyPayload {
  sourceChannel: number;
  sourceStartCol: number;
  sourceEndCol: number;
  destChannel: number;
  destCol: number;
}

export interface GridPastePayload {
  sourceChannel: number;
  sourceStartCol: number;
  destChannel: number;
  destCol: number;
  clearSource: boolean;
}

// Pattern Commands
export interface PatternDto {
  id: number;
  name: string;
  length: number;
  grid: NoteDto[][];
}

export interface NoteDto {
  pitch: number;
  velocity: number;
  start: number;
  duration: number;
}

export interface PatternUpdateRequest {
  channel: number;
  patternIndex: number;
  cellRow: number;
  cellCol: number;
  note: NoteDto | null;
}

// Channel Commands
export interface ChannelAddPayload {
  name?: string;
  instrumentId?: string;
}

export interface ChannelRemovePayload {
  channel: number;
}

export interface ChannelSetNamePayload {
  channel: number;
  name: string;
}

export interface ChannelSetVolumePayload {
  channel: number;
  volume: number; // 0.0 - 1.0
}

export interface ChannelSetPanPayload {
  channel: number;
  pan: number; // -1.0 (left) to 1.0 (right)
}

export interface ChannelSetMutePayload {
  channel: number;
  muted: boolean;
}

export interface ChannelSetSoloPayload {
  channel: number;
  solo: boolean;
}

// Instrument Commands
export interface InstrumentListPayload {
  // No payload - returns all available instruments
}

export interface InstrumentLoadPayload {
  path: string;
}

export interface InstrumentAssignPayload {
  channel: number;
  instrumentId: string;
}

// Preview Commands
export interface NotePreviewPayload {
  channel: number;
  note: number;
  velocity: number;
}

export interface NotePreviewResponse {
  success: boolean;
}

// ==================== SOURCE COMMANDS ====================

export interface ListSourcesResponse {
  sources: SourceInfo[];
}

export interface SourceInfo {
  id: string;
  name: string;
  category: string;
}

export interface ChannelSetSourcePayload {
  channel: number;
  source_id: string;
}

export interface ChannelSetSourceResponse {
  channel: number;
  source_id: string;
  success: boolean;
}

export interface GetSourceParamsPayload {
  source_id: string;
}

export interface ParamInfo {
  id: string;
  name: string;
  min: number;
  max: number;
  default: number;
  unit: string;
}

export interface GetSourceParamsResponse {
  source_id: string;
  params: ParamInfo[];
}

export interface SetSourceParamPayload {
  channel: number;
  param_id: string;
  value: number;
}

export interface SetSourceParamResponse {
  channel: number;
  param_id: string;
  value: number;
  success: boolean;
}

// ==================== COMMAND RESPONSE TYPES ====================

export interface ProjectNewResponse {
  projectId: string;
  name: string;
  createdAt: string;
}

export interface ProjectSaveResponse {
  path: string;
  savedAt: string;
}

export interface ProjectLoadResponse {
  projectId: string;
  name: string;
  path: string;
  bpm: number;
  timeSignature: { numerator: number; denominator: number };
  channels: ChannelState[];
}

export interface ProjectExportResponse {
  path: string;
  exportedAt: string;
  duration: number;
}

export interface TransportPlayResponse {
  playing: boolean;
  position: number;
}

export interface TransportStopResponse {
  playing: boolean;
  position: number;
}

export interface TransportPauseResponse {
  playing: boolean;
}

export interface TransportSeekResponse {
  position: number;
}

export interface TransportSetBpmResponse {
  bpm: number;
}

export interface TransportSetTimeSignatureResponse {
  numerator: number;
  denominator: number;
}

export interface GridSetCellResponse {
  success: boolean;
}

export interface GridClearCellResponse {
  success: boolean;
}

export interface GridClearChannelResponse {
  success: boolean;
}

export interface GridClearAllResponse {
  success: boolean;
}

export interface GridCopyResponse {
  success: boolean;
  cellsCopied: number;
}

export interface GridPasteResponse {
  success: boolean;
  cellsPasted: number;
}

export interface PatternUpdateResponse {
  success: boolean;
}

export interface ChannelAddResponse {
  channel: number;
  name: string;
}

export interface ChannelRemoveResponse {
  success: boolean;
}

export interface ChannelSetNameResponse {
  channel: number;
  name: string;
}

export interface ChannelSetVolumeResponse {
  channel: number;
  volume: number;
}

export interface ChannelSetPanResponse {
  channel: number;
  pan: number;
}

export interface ChannelSetMuteResponse {
  channel: number;
  muted: boolean;
}

export interface ChannelSetSoloResponse {
  channel: number;
  solo: boolean;
}

export interface InstrumentListResponse {
  instruments: InstrumentInfo[];
}

export interface InstrumentLoadResponse {
  instrumentId: string;
  name: string;
}

export interface InstrumentAssignResponse {
  channel: number;
  instrumentId: string;
}

// ==================== EVENT PAYLOADS ====================

export interface TransportStateEventPayload {
  playing: boolean;
  position: number; // Current position in beats
  bpm: number;
  timeSignature: { numerator: number; denominator: number };
}

export interface GridChangeEventPayload {
  channel: number;
  row: number;
  col: number;
  velocity: number;
  enabled: boolean;
}

export interface ChannelChangeEventPayload {
  channel: number;
  name: string;
  volume: number;
  pan: number;
  muted: boolean;
  solo: boolean;
}

export interface ProjectChangedEventPayload {
  projectId: string;
  name: string;
  path: string | null;
  modified: boolean;
}

export interface AudioEngineStateEventPayload {
  initialized: boolean;
  sampleRate: number;
  bufferSize: number;
}

export interface ErrorEventPayload {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

export interface PlayheadEventPayload {
  bar: number;
  beat: number;
  step: number;
  tick: number;
  is_playing: boolean;
}

export interface LevelsEventPayload {
  channels: number[];
  master: number;
}

// ==================== SHARED TYPES ====================

export interface ChannelState {
  channel: number;
  name: string;
  volume: number;
  pan: number;
  muted: boolean;
  solo: boolean;
  instrumentId: string | null;
}

export interface InstrumentInfo {
  id: string;
  name: string;
  category: string;
}

export interface GridCell {
  channel: number;
  row: number;
  col: number;
  velocity: number;
  enabled: boolean;
}

export interface GridState {
  channels: number;
  rows: number;
  cols: number;
  cells: GridCell[];
}

File: src-tauri/src/dto.rs
// dto.rs
//
// Purpose: Backend DTOs for Tauri command payloads
//
// Request/response structs for project commands.

use serde::{Deserialize, Serialize};

use crate::project::{Note, Pattern, Project};

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectNewRequest {
    pub name: Option<String>,
}

impl Default for ProjectNewRequest {
    fn default() -> Self {
        Self { name: None }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectNewResponse {
    pub project: Project,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectSaveRequest {
    pub project: Project,
    pub path: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectSaveResponse {
    pub success: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectLoadRequest {
    pub path: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectLoadResponse {
    pub project: Project,
}

// Pattern Update DTOs

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct NoteDto {
    pub pitch: u8,
    pub velocity: u8,
    pub start: u16,
    pub duration: u16,
}

impl From<Note> for NoteDto {
    fn from(note: Note) -> Self {
        Self {
            pitch: note.pitch,
            velocity: note.velocity,
            start: note.start,
            duration: note.duration,
        }
    }
}

impl From<NoteDto> for Note {
    fn from(dto: NoteDto) -> Self {
        Self {
            pitch: dto.pitch,
            velocity: dto.velocity,
            start: dto.start,
            duration: dto.duration,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PatternUpdateRequest {
    pub channel: usize,
    pub pattern_index: usize,
    pub cell_row: usize,
    pub cell_col: usize,
    pub note: Option<NoteDto>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PatternUpdateResponse {
    pub success: bool,
}

// Transport DTOs

#[derive(Debug, Serialize, Deserialize)]
pub struct TransportPlayRequest {
    pub position: Option<u32>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TransportPlayResponse {
    pub playing: bool,
    pub position: u32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TransportStopResponse {
    pub playing: bool,
    pub position: u32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TransportSetTempoRequest {
    pub tempo: u16,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TransportSetTempoResponse {
    pub tempo: u16,
}

// Event Payloads for frontend synchronization

/// Playhead position event - emitted on each step change during playback
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayheadEvent {
    /// Current bar number (0-indexed)
    pub bar: u32,
    /// Current beat within the bar (0-3)
    pub beat: u32,
    /// Current step within the beat (0-15, 16th notes)
    pub step: u32,
    /// Current tick within the step (0-23)
    pub tick: u32,
    /// Whether playback is active
    pub is_playing: bool,
}

/// Audio levels event - peak levels per channel
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LevelsEvent {
    /// Peak level for each channel (0.0 to 1.0+)
    pub channels: Vec<f32>,
    /// Master peak level
    pub master: f32,
}

// Source Registry DTOs

#[derive(Debug, Serialize, Deserialize)]
pub struct SourceInfo {
    pub id: String,
    pub name: String,
    pub category: String,
}

impl From<crate::plugin::SourceDescriptor> for SourceInfo {
    fn from(desc: crate::plugin::SourceDescriptor) -> Self {
        Self {
            id: desc.id,
            name: desc.name,
            category: desc.category.as_str().to_string(),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ListSourcesResponse {
    pub sources: Vec<SourceInfo>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ChannelSetSourceRequest {
    pub channel: usize,
    pub source_id: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ChannelSetSourceResponse {
    pub channel: usize,
    pub source_id: String,
    pub success: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GetSourceParamsRequest {
    pub source_id: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ParamInfo {
    pub id: String,
    pub name: String,
    pub min: f32,
    pub max: f32,
    pub default: f32,
    pub unit: String,
}

impl From<crate::plugin::ParamDescriptor> for ParamInfo {
    fn from(param: crate::plugin::ParamDescriptor) -> Self {
        Self {
            id: param.id.0,
            name: param.name,
            min: param.min,
            max: param.max,
            default: param.default,
            unit: param.unit,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GetSourceParamsResponse {
    pub source_id: String,
    pub params: Vec<ParamInfo>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SetSourceParamRequest {
    pub channel: usize,
    pub param_id: String,
    pub value: f32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SetSourceParamResponse {
    pub channel: usize,
    pub param_id: String,
    pub value: f32,
    pub success: bool,
}
